crates/batcher/src/lib.rs

```diff
@@ -491,6 +500,243 @@ impl Batcher {
         .map_err(|e| e.inner())
     }
 
+    /// Poll for BalanceUnlocked events from BatcherPaymentService contract.
+    /// Runs at configurable intervals and checks recent blocks for events (2x the polling interval).
+    /// When an event is detected, removes user's proofs from queue and resets UserState.
+    pub async fn poll_balance_unlocked_events(self: Arc<Self>) -> Result<(), BatcherError> {
+        let mut interval = tokio::time::interval(tokio::time::Duration::from_secs(
+            self.balance_unlock_polling_interval_seconds,
+        ));
+
+        loop {
+            interval.tick().await;
+
+            if let Err(e) = self.process_balance_unlocked_events().await {
+                error!("Error processing BalanceUnlocked events: {:?}", e);
+                // Continue polling even if there's an error
+            }
+        }
+    }
+
+    async fn process_balance_unlocked_events(&self) -> Result<(), BatcherError> {
+        // Get current block number using HTTP providers
+        let current_block = match self.get_current_block_number().await {
+            Ok(block) => block,
+            Err(e) => {
+                warn!("Failed to get current block number: {:?}", e);
+                return Ok(());
+            }
+        };
+
+        // Calculate the block range based on polling interval
+        // Formula: interval / 12 * 2 (assuming 12-second block times, look back 2x the interval)
+        let block_range = (self.balance_unlock_polling_interval_seconds / 12) * 2;
+        let from_block = current_block.saturating_sub(U64::from(block_range));
+
+        // Query events with retry logic
+        let events = match self
+            .query_balance_unlocked_events(from_block, current_block)
+            .await
+        {
+            Ok(events) => events,
+            Err(e) => {
+                warn!(
+                    "Failed to query BalanceUnlocked events after retries: {:?}",
+                    e
+                );
+                return Ok(());
+            }
+        };
+
+        info!(
+            "Found {} BalanceUnlocked events in blocks {} to {}",
+            events.len(),
+            from_block,
+            current_block
+        );
+
+        // Process each event
+        for event in events {
+            let user_address = event.user;
+            info!(
+                "Processing BalanceUnlocked event for user: {:?}",
+                user_address
+            );
+
+            // Check if user has proofs in queue
+            if self.user_has_proofs_in_queue(user_address).await {
+                info!(
+                    "User {:?} has proofs in queue, verifying funds are still unlocked",
+                    user_address
+                );
+
+                // Double-check that funds are still unlocked by calling the contract
```

Comment:
Not necessary and it might be obvious, but a better explanation of why we are doing this:

```suggestion
                // Double-check that funds are still unlocked by calling the contract
               // This is necessary because we query events over a block range, and the
              // user’s state may have changed (e.g., funds could be locked again) after
             // the event was emitted. Verifying on-chain ensures we don’t act on stale data.
```